src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs

```diff
@@ -191,17 +207,25 @@ void Connect()
 
             private void Dispose()
             {
-                if (_disposed)
+                lock (_gate)
                 {
-                    Debug.Fail("Tagger already disposed");
-                    return;
-                }
+                    // Perform this all logic related to _disposed while holding _gate so that we get
+                    // a consistent view of the world in EnqueueWork.
 
-                // Stop computing any initial tags if we've been asked for them.
-                _disposalTokenSource.Cancel();
-                _cancellationSeries.Dispose();
+                    if (_disposed)
+                    {
+                        Debug.Fail("Tagger already disposed");
+                        return;
+                    }
+
+                    // Stop computing any initial tags if we've been asked for them.  
+                    _disposalTokenSource.Cancel();
```

Comment:
Use `CancelAfter` to avoid performing callbacks while holding the lock

```suggestion
                    _disposalTokenSource.CancelAfter(0);
```